### PR TITLE
Add Optional Model Quantization for Backend Performance Optimization

### LINKED ISSUE #554: [Enhancement] Add Optional Model Quantization to Reduce Memory Usage and Improve Inference Performance
### Feature Description

The EduAid backend currently loads several large NLP models in full FP32 precision, which results in significant memory consumption and increased inference latency.

Some of the major models currently used include:

- T5-large for MCQ generation
- T5-base for question generation
- DistilBERT for NLI-based boolean answer prediction
- SpaCy for NLP processing
- Sense2Vec for distractor generation

These models are currently loaded in full precision which increases:

- Memory usage
- Server resource consumption
- Cold start latency
- Risk of Out-of-Memory (OOM) errors when handling multiple requests

### Proposed Enhancement

Introduce **optional model quantization and precision optimization** during model loading to reduce memory footprint and improve inference speed.

This can be implemented inside:

```
backend/Generator/main.py
```

Specifically in the:

```
ModelManager.load_model()
```

method.

### Possible Implementation

1. Apply **dynamic INT8 quantization** for transformer models when running on CPU.

Example:

```python
torch.quantization.quantize_dynamic(
    model,
    {torch.nn.Linear},
    dtype=torch.qint8
)
```

2. Use **FP16 precision when CUDA is available**.

Example:

```python
model = model.half()
```

3. Add a configuration flag such as:

```
ENABLE_MODEL_QUANTIZATION = True
```

to allow enabling or disabling this behavior.

### Expected Benefits

- Reduce memory consumption by ~40–60%
- Faster inference for transformer models
- Lower server resource usage
- Better scalability for concurrent requests
- Reduced deployment costs

### Why This Matters

Currently the backend may load multiple large models simultaneously which can lead to memory usage exceeding **5–7GB** in some environments.

Applying quantization could reduce this significantly while maintaining acceptable model accuracy for question generation tasks.

### Files Likely Affected

```
backend/Generator/main.py
backend/server.py
```

### Additional Context

This improvement focuses on optimizing model resource usage without changing the existing functionality of the quiz generation pipeline.

The change would be fully backward compatible and could be implemented behind a feature flag so that deployments can enable or disable quantization depending on the available hardware.

Models that would benefit most from this improvement include:

- T5-large models used in MCQ generation
- T5-base models used for question generation
- DistilBERT-based NLI models used for boolean answer prediction

Other models such as SpaCy and Sense2Vec may not significantly benefit from quantization but can still remain part of the existing pipeline.

This enhancement could significantly improve backend performance and make the system more efficient for production deployments.

### Code of Conduct

- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and will post updates there
- [x] I have searched existing issues to avoid duplicates

### PR DESCRIPTION
**Pull request by piyush06singhal**
### Addressed Issues:

Fixes #554 

### Additional Notes:

This PR introduces optional model quantization support to reduce memory usage and improve inference performance for the EduAid backend models. The feature is disabled by default and can be enabled through environment variables.

Key points:
- Adds optional quantization for transformer models
- Supports INT8 dynamic quantization for CPU
- Supports FP16 half precision for GPU
- Fully backward compatible
- No changes to existing APIs or workflows

## Description

The EduAid backend loads several large transformer models (T5-large, T5-base, DistilBERT, BERT) using FP32 precision, which leads to high memory consumption and increased computational load.

This PR introduces an optional quantization mechanism to reduce memory usage and improve inference performance. Quantization is applied immediately after model loading and device placement without modifying the existing architecture.

### Modified Files

`backend/Generator/main.py`

- Added configuration import with fallback handling
- Implemented `apply_quantization()` helper function
- Applied quantization to transformer models after loading

### New File

`backend/config.py`

- Configuration module for quantization settings
- Reads configuration values from environment variables
- Provides default safe values

## Supported Quantization Modes

INT8 (CPU)

- Uses PyTorch dynamic quantization

FP16 (GPU)

- Uses half precision inference

Quantization is only applied when explicitly enabled.

## Usage

Enable INT8 (CPU)

```bash
export ENABLE_MODEL_QUANTIZATION=true
export MODEL_PRECISION=int8
python server.py
```

Enable FP16 (GPU)

```bash
export ENABLE_MODEL_QUANTIZATION=true
export MODEL_PRECISION=fp16
python server.py
```

Default behavior (quantization disabled)

```bash
python server.py
```

## Testing

Verified that:

- Backend runs normally when quantization is disabled
- INT8 quantization works correctly on CPU
- FP16 quantization works correctly on GPU
- All endpoints function as expected
- Model outputs remain valid
- Quantization only activates when enabled
- System gracefully falls back to FP32 if quantization fails

## Breaking Changes

None.  
This change is fully backward compatible.

## Checklist

- [x] My PR addresses a single issue, fixes a single bug or makes a single improvement.
- [x] My code follows the project's code style and conventions
- [ ] If applicable, I have made corresponding changes or additions to the documentation
- [ ] If applicable, I have made corresponding changes or additions to tests
- [x] My changes generate no new warnings or errors
- [x] I have joined the [Discord server](https://discord.gg/hjUhu33uAn) and I will share a link to this PR with the project maintainers there
- [x] I have read the [Contribution Guidelines](../CONTRIBUTING.md)
- [x] Once I submit my PR, CodeRabbit AI will automatically review it and I will address CodeRabbit's comments.
- [x] I have filled this PR template completely and carefully, and I understand that my PR may be closed without review otherwise.
---

Check one of the checkboxes below:

- [ ] This PR does not contain AI-generated code at all.
- [x] This PR contains AI-generated code. I have read the [AI Usage Policy](https://github.com/AOSSIE-Org/Info/blob/main/AI-UsagePolicy.md) and this PR complies with this policy. I have tested the code locally and I am responsible for it.

I have used the following AI models and tools: TODO

### AI Usage Disclosure:

We encourage contributors to use AI tools responsibly when creating Pull Requests. While AI can be a valuable aid, it is essential to ensure that your contributions meet the task requirements, build successfully, include relevant tests, and pass all linters. Submissions that do not meet these standards may be closed without warning to maintain the quality and integrity of the project. Please take the time to understand the changes you are proposing and their impact. AI slop is strongly discouraged and may lead to banning and blocking. Do not spam our repos with AI slop.

## Summary by CodeRabbit

* **New Features**
  * Configurable, device-aware model quantization (precision control and safe fallbacks)
  * Outputs now include timing metadata for question-generation flows

* **Bug Fixes**
  * Ensures model inputs run on the correct device and cleans CUDA buffers after quantized inference

* **Chores**
  * Added lightweight runtime logging for quantization actions and decisions

### CODE REVIEW

**Comment by piyush06singhal** [comment]
@coderabbitai review
